File: conformance/work-class/1.0.0-draft.2/typescript/evidence.ts
import { AdmissionError, byteCompare, canonical, digest } from './canonical.ts';
import { requireShape, shape, VERSION } from './schema.ts';
import { same } from './values.ts';
import type { RecordValue as Row } from './values.ts';
import { validateWorkClass } from './work-class.ts';

const clone = <T>(value: T): T => structuredClone(value);
const fail = (code: string, path = ''): never => { throw new AdmissionError(code, path); };

export function readback(value: unknown, path = ''): { path: string; value: string }[] {
  if (value === null || typeof value !== 'object') return [{ path, value: canonical(value) }];
  const entries = Object.entries(value as object);
  if (!entries.length) return [{ path, value: canonical(value) }];
  return entries.flatMap(([key, child]) => readback(child, path + '/' + key.replaceAll('~', '~0').replaceAll('/', '~1')))
    .sort((a, b) => byteCompare(a.path, b.path));
}

function failure(request: Row, code: string): Row {
  return { status: 'EVIDENCE_INVALID', profile: request.profile, role: request.role, code, path: '', specification_pin: request.specification_pin };
}

function sorted(values: string[]): boolean { return values.every((value, i) => i === 0 || byteCompare(values[i - 1], value) < 0); }

// The correspondence form of `check-evidence` after the shared subject and status checks; `deploy`
// applies the same conditions to the correspondence record named by the deployment authorization.
export function correspondenceCode(evidence: Row, workClass: Row, workDigest: string): string | null {
  if (evidence.work_class_digest !== workDigest || evidence.source_id !== workClass.source.id || evidence.source_revision !== workClass.source.revision) return 'BINDING_MISMATCH';
  const mappings = evidence.mappings as Row[];
  if (!sorted(mappings.map(row => row.source_obligation_id)) || mappings.some(row => !sorted(row.provision_paths))) return 'BINDING_MISMATCH';
  if (!sorted(evidence.uncovered_source_obligation_ids) || !sorted(evidence.unsupported_provision_paths)) return 'BINDING_MISMATCH';
  const mappedIds = mappings.map(row => row.source_obligation_id), allSource = [...mappedIds, ...evidence.uncovered_source_obligation_ids].sort(byteCompare);
  if (new Set(allSource).size !== allSource.length || !same(allSource, workClass.source.obligations)) return 'BINDING_MISMATCH';
  const universe = readback(workClass).map(row => row.path);
  const mappedPaths = mappings.flatMap(row => row.provision_paths);
  const coveredPaths = [...new Set([...mappedPaths, ...evidence.unsupported_provision_paths])].sort(byteCompare);
  if (mappedPaths.some((path: string) => evidence.unsupported_provision_paths.includes(path)) || !same(coveredPaths, universe)) return 'BINDING_MISMATCH';
  const edges = mappings.flatMap(row => row.provision_paths.map((path: string) => canonical([row.source_obligation_id, path])));
  if (new Set(edges).size !== edges.length) return 'BINDING_MISMATCH';
  const residue = evidence.residue as Row[];
  const residueKeys = residue.map(row => canonical([row.kind, row.subject]));
  if (new Set(residueKeys).size !== residueKeys.length || !sorted(residueKeys)) return 'BINDING_MISMATCH';
  const expectedResidue = [
    ...evidence.uncovered_source_obligation_ids.map((value: string) => canonical(['SOURCE_OBLIGATION', value])),
    ...evidence.unsupported_provision_paths.map((value: string) => canonical(['ARTIFACT_PROVISION', value])),
  ].sort(byteCompare);
  if (!same(residueKeys, expectedResidue)) return 'BINDING_MISMATCH';
  const reviewSubject = clone(evidence); delete reviewSubject.review;
  if (evidence.review.subject_digest !== digest('correspondence-subject', reviewSubject)) return 'BINDING_MISMATCH';
  if (evidence.review.status !== 'ACCEPTED') return 'EVIDENCE_NOT_ACCEPTED';
  return null;
}

export function checkEvidence(input: unknown, selectedPin: string): Row {
  if (!shape('evidence.schema.json', 'input', input)) return { status: 'REFUSED', code: 'SCHEMA_INVALID', path: '' };
  const request = input as Row;
  if (request.specification_pin !== selectedPin) return { status: 'REFUSED', code: 'VERSION_UNSUPPORTED', path: '' };
  if (request.role !== 'REVIEW_EVIDENCE') return failure(request, 'BINDING_MISMATCH');
  const subject = request.subject, evidence = request.evidence;
  if (subject.candidate_identity !== VERSION || subject.specification_pin !== selectedPin || evidence.subject_digest !== subject.subject_digest) return failure(request, 'BINDING_MISMATCH');
  if (evidence.status !== 'ACCEPTED') return failure(request, 'EVIDENCE_NOT_ACCEPTED');

  if (evidence.kind === 'CORRESPONDENCE') {
    let workDigest: string;
    try { workDigest = validateWorkClass(request.work_class, selectedPin).digest; }
    catch { return failure(request, 'BINDING_MISMATCH'); }
    if (request.profile !== request.work_class.profile || subject.kind !== 'WORK_CLASS' || subject.subject_id !== request.work_class.id || request.work_class_digest !== workDigest || subject.subject_digest !== workDigest) return failure(request, 'BINDING_MISMATCH');
    const code = correspondenceCode(evidence, request.work_class, workDigest);
    if (code) return failure(request, code);
  }
  return { status: 'EVIDENCE_VALID', profile: request.profile, role: request.role, subject_digest: subject.subject_digest, evidence_digest: digest('evidence', evidence), specification_pin: selectedPin };
}

export function validateEvidenceArtifact(value: unknown): string {
  requireShape('evidence.schema.json', 'record', value);
  return digest('evidence', value);
}

export function validateBoundary(value: unknown, selectedPin: string): string {
  requireShape('boundary.schema.json', 'record', value);
  const record = value as Row;
  if (record.candidate_identity !== VERSION || record.specification_pin !== selectedPin) fail('VERSION_UNSUPPORTED');
  if (!sorted(record.paragraph_ids) || !sorted(record.requirement_ids)) fail('INPUT_INVALID');
  if (record.establishes === record.does_not_establish) fail('INPUT_INVALID');
  return digest('boundary', record);
}

File: conformance/work-class/1.0.0-draft.2/python/implementation/evidence.py
"""Draft-2 evidence and correspondence checks."""

from __future__ import annotations

from .common import PIN, digest, readback_lines, schema_valid, utf8_sorted_unique, valid_instant


def _invalid(request, code):
    return {
        "status": "EVIDENCE_INVALID",
        "profile": request["profile"],
        "role": request["role"],
        "code": code,
        "path": "",
        "specification_pin": PIN,
    }


def _valid(request):
    return {
        "status": "EVIDENCE_VALID",
        "profile": request["profile"],
        "role": request["role"],
        "subject_digest": request["subject"]["subject_digest"],
        "evidence_digest": digest("evidence", request["evidence"]),
        "specification_pin": PIN,
    }


def _basic(request):
    subject = request["subject"]
    evidence = request["evidence"]
    if request["specification_pin"] != PIN or subject["specification_pin"] != PIN:
        return _invalid(request, "BINDING_MISMATCH")
    if subject["candidate_identity"] != "seampoint.work-class/1.0.0-draft.2":
        return _invalid(request, "BINDING_MISMATCH")
    if evidence["subject_digest"] != subject["subject_digest"]:
        return _invalid(request, "BINDING_MISMATCH")
    if not valid_instant(evidence["recorded_at"]) or (evidence.get("expires_at") is not None and not valid_instant(evidence["expires_at"])):
        return _invalid(request, "SCHEMA_INVALID")
    if evidence["status"] != "ACCEPTED":
        return _invalid(request, "EVIDENCE_NOT_ACCEPTED")
    for key in ("scope_ids", "decision_ids", "authorization_scope", "paragraph_ids", "case_ids", "finding_ids"):
        if key in evidence and not utf8_sorted_unique(evidence[key]):
            return _invalid(request, "BINDING_MISMATCH")
    return _valid(request)


def _correspondence(request):
    subject = request["subject"]
    work_class = request["work_class"]
    correspondence = request["evidence"]
    if not valid_instant(correspondence["recorded_at"]) or not valid_instant(correspondence["review"]["recorded_at"]):
        return _invalid(request, "SCHEMA_INVALID")
    work_digest = digest("work-class-definition", work_class)
    if request["specification_pin"] != PIN or subject["specification_pin"] != PIN:
        return _invalid(request, "BINDING_MISMATCH")
    if work_class["specification_pin"] != PIN or request["profile"] != work_class["profile"]:
        return _invalid(request, "BINDING_MISMATCH")
    if subject["kind"] != "WORK_CLASS" or subject["subject_id"] != work_class["id"]:
        return _invalid(request, "BINDING_MISMATCH")
    if any(value != work_digest for value in (
        request["work_class_digest"], subject["subject_digest"],
        correspondence["subject_digest"], correspondence["work_class_digest"],
    )):
        return _invalid(request, "BINDING_MISMATCH")
    if correspondence["source_id"] != work_class["source"]["id"] or correspondence["source_revision"] != work_class["source"]["revision"]:
        return _invalid(request, "BINDING_MISMATCH")

    obligations = work_class["source"]["obligations"]
    mappings = correspondence["mappings"]
    uncovered = correspondence["uncovered_source_obligation_ids"]
    unsupported = correspondence["unsupported_provision_paths"]
    paths = [row["path"] for row in readback_lines(work_class)]
    mapping_ids = [row["source_obligation_id"] for row in mappings]
    mapped_paths = [item for row in mappings for item in row["provision_paths"]]
    if not utf8_sorted_unique(mapping_ids) or any(not utf8_sorted_unique(row["provision_paths"]) for row in mappings):
        return _invalid(request, "BINDING_MISMATCH")
    if not utf8_sorted_unique(uncovered) or sorted(mapping_ids + uncovered, key=lambda value: value.encode("utf-8")) != obligations:
        return _invalid(request, "BINDING_MISMATCH")
    if not utf8_sorted_unique(unsupported) or set(mapped_paths) & set(unsupported):
        return _invalid(request, "BINDING_MISMATCH")
    if set(mapped_paths) | set(unsupported) != set(paths):
        return _invalid(request, "BINDING_MISMATCH")
    if len([(row["source_obligation_id"], path) for row in mappings for path in row["provision_paths"]]) != len(set((row["source_obligation_id"], path) for row in mappings for path in row["provision_paths"])):
        return _invalid(request, "BINDING_MISMATCH")

    residues = correspondence["residue"]
    expected_residues = {("SOURCE_OBLIGATION", item) for item in uncovered} | {("ARTIFACT_PROVISION", item) for item in unsupported}
    actual_residues = [(row["kind"], row["subject"]) for row in residues]
    if len(actual_residues) != len(set(actual_residues)) or set(actual_residues) != expected_residues:
        return _invalid(request, "BINDING_MISMATCH")
    if actual_residues != sorted(actual_residues, key=lambda row: (row[0].encode("utf-8"), row[1].encode("utf-8"))):
        return _invalid(request, "BINDING_MISMATCH")

    if correspondence["status"] != "ACCEPTED":
        return _invalid(request, "EVIDENCE_NOT_ACCEPTED")
    review = correspondence["review"]
    subject_body = dict(correspondence)
    del subject_body["review"]
    if review["subject_digest"] != digest("correspondence-subject", subject_body):
        return _invalid(request, "BINDING_MISMATCH")
    if review["status"] != "ACCEPTED":
        return _invalid(request, "EVIDENCE_NOT_ACCEPTED")
    return _valid(request)


def check(request):
    if not schema_valid(request, "evidence.schema.json", "input"):
        profile = request.get("profile") if isinstance(request, dict) else None
        role = request.get("role") if isinstance(request, dict) else None
        if "specification_pin" in request and profile in {"LINEAR", "CHOICE_LOOPS", "DEADLINES", "PARALLEL_FANOUT"} and role == "REVIEW_EVIDENCE":
            return _invalid(request, "SCHEMA_INVALID")
        return {"status": "REFUSED", "code": "SCHEMA_INVALID", "path": ""}
    if request["specification_pin"] != PIN:
        return {"status":"REFUSED","code":"VERSION_UNSUPPORTED","path":""}
    if "work_class" in request:
        return _correspondence(request)
    return _basic(request)
